Generated columns

escopecz opened the pull request:
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR is adding a [generated column](https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html) based on `email_stats.date_sent`. It will generate the value in format `Y-m-d` and build a new index from it. Then it modifies the line chart queries to use that column instead of formatting the original date.

Original query to load line chart data from email stats table:
```
SELECT DATE_FORMAT(t.date_sent, '%Y-%m-%d') AS date, COUNT(*) AS count FROM mautic_email_stats t WHERE (t.email_id IN (3)) AND (t.date_sent BETWEEN :dateFrom AND :dateTo) GROUP BY DATE_FORMAT(t.date_sent, '%Y-%m-%d') ORDER BY DATE_FORMAT(t.date_sent, '%Y-%m-%d') ASC LIMIT 31
```

New query based on the generated column:
```
SELECT t.generated_sent_date AS date, COUNT(*) AS count FROM mautic_email_stats t WHERE (t.email_id IN (3)) AND (t.date_sent BETWEEN :dateFrom AND :dateTo) GROUP BY t.generated_sent_date ORDER BY t.generated_sent_date ASC LIMIT 31
```

The speed improvement for the select query is somewhere around 3,5x faster. To be more specific, running the query on that 6.6M row table took ~700ms. With this PR it loads in ~200ms.

I've tested adding the new generated column on a email_stats table with 6.6M rows and it took 172 seconds to add it. So it should not cause any troubles to migrate even big instances. But do not use the UI updater. Use the script!

This is an experiment. If we find that it's not causing any troubles we should implement it also for date_read and other date columns in other stat tables used in the line charts.

#### Steps to test this PR:
1. Ensure you can see the generated column definition when viewing the schema changes. You should see something like this:
```
$ app/console doctrine:schema:update --dump-sql | grep generated
ALTER TABLE mautic_email_stats ADD generated_sent_date DATE AS (CONCAT(YEAR(date_sent), "-", LPAD(MONTH(date_sent), 2, "0"), "-", LPAD(DAY(date_sent), 2, "0"))) COMMENT '(DC2Type:generated)';
CREATE INDEX mautic_generated_sent_date_email_id ON mautic_email_stats (generated_sent_date, email_id);
```
2. Run the migration. Generated fields doesn't have a migration file because we don't know whether the database supports generated fields. So we check the database version and run the alter table queries only if the database supports it. You should see this during migration:
```
$ app/console doctrine:migration:migrate

++ adding generated column generated_sent_date
-> ALTER TABLE mautic_email_stats
            ADD generated_sent_date DATE AS (CONCAT(YEAR(date_sent), "-", LPAD(MONTH(date_sent), 2, "0"), "-", LPAD(DAY(date_sent), 2, "0"))) COMMENT '(DC2Type:generated)',
            ADD index `mautic_generated_sent_date_email_id`(generated_sent_date, email_id)
++ generated column added (generated columns: 60.00 MiB - 36 ms)
                    Mautic Migrations
WARNING! You are about to execute a database migration that could result in schema changes and data lost. Are you sure you wish to continue? (y/n)No migrations to execute.
```

3. Load a segment email detail page, report based on email stats with the line chart and dashboard with the emails sent widget. All should load correctly (hopefully faster than before) and it should show the same values as before. Test also the sent email dashboard widget and report widget. All time units should still work.

escopecz committed:
New Generated column type and use it for generated_sent_date

escopecz committed:
Adding generated field to the ChartQuery - WIP

escopecz committed:
ChartQuery generated column feature tested

escopecz committed:
date_sent generated column used also when loading report email stat chart

escopecz committed:
DATE type is 2 times faster than CHAR(10)

escopecz committed:
Remove debug

escopecz committed:
Property comments added

escopecz committed:
Forgotten rename fixed

alanhartless committed:
Doctrine would try to persist a value for a generated column when persisting an entity which MySQL would reject. This bypasses that by defining the column and index in a subscriber instead of on the entity itself.

alanhartless committed:
Fixed migration to match updated index name and Doctrine type

alanhartless committed:
Removed unused platformOptions (previously required for GeneratedType when auto-generating columnDefinitions) and added notNull => false to prevent false positive changes when using `doctrine:schema:update`

escopecz committed:
Generated columns are gathered via event so they won't be defined on 3 places

escopecz committed:
Fixing tests

escopecz committed:
Create generated columns automatically before the migration command is run

escopecz committed:
Formating fix

escopecz committed:
First batch of tests

escopecz committed:
Second batch of tests

escopecz committed:
Third batch of tests

escopecz committed:
DB version is loaded from the config if set there

escopecz committed:
db_server_version cannot be missing. Check for the default value instead

escopecz committed:
Tests fixed

escopecz committed:
Split the generated column migration alter table queries into 2 for safer execution.

escopecz committed:
Tests fixed after the query modification

escopecz committed:
MySql supports generated columns with index since 5.7.14

escopecz committed:
Wrong variable name fixed

escopecz committed:
Remove the public keyword as it's PHP 7.1+

escopecz committed:
Removing private keyword before const to make the code compatible with PHP versions before 7.1

npracht commented:
Travis my love...

escopecz commented:
Honestly, I don't know what to do with Travis in this case. It passes all tests on my local. Maybe a database version issue? I'd appreciate any idea how to fix it.

RCheesley commented:
@escopecz looking like we've got some conflicts with this one too - presume it's going to need refactoring for potential inclusion with 3.1

npracht commented:
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

npracht commented:
Hi @escopecz is this still relevant for M3 ?
Thanks !

escopecz commented:
Probably

escopecz committed:
CS Fixer changes

escopecz committed:
M3 updates

escopecz committed:
Namespace fix

escopecz committed:
oops, implements instead of extends an interface

escopecz committed:
Adding back event that was lost during rebase

escopecz committed:
Test fix

escopecz committed:
Return early, return types added part 1

escopecz committed:
PHPSTAN fixes

escopecz committed:
We cannot create the index on the entity as it causes...

"There is no column with name 'generated_sent_date' on table 'email_stats'" error on schema generate

escopecz committed:
A test fix

escopecz committed:
unit => Unit folder rename

escopecz committed:
PHPSTAN fixes

escopecz committed:
Adding types as this will be merged to M3

escopecz committed:
originalDateColumn can be null

escopecz committed:
Forgot to update the interface

escopecz commented:
FYI there are some critical fixes that we cannot provide to the community until this will be merged.

npracht commented:
Then this should be a priority for 3.1.0

dennisameling commented on app/bundles/CoreBundle/Doctrine/Provider/GeneratedColumnsProvider.php:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @see https://mariadb.com/kb/en/library/generated-columns
+     */
+    const MARIADB_MINIMUM_VERSION = '5.2';
```
Shouldn't this be 10.2 (see https://mariadb.com/kb/en/generated-columns)?

> In MariaDB 10.2 and later, the syntax is also compatible with the syntax for MySQL's generated columns.

dennisameling reviewed (changes requested):
Code LGTM, just left a comment about the minimal MariaDB version.

Note for 2nd reviewer: **this PR can NOT be reviewed on Mautibox because the DB migration needs to be applied manually.**

@escopecz you're right about the DB version check. While we do check for the DB version in the M2 > M3 upgrade script, we don't check it during installation. Added that to https://github.com/mautic/mautic/issues/8452 for later consideration. One minor thing is that the "[Community Release Leader](https://docs.google.com/document/d/11psNpVYpcn4m9ZxyD_a5panpK_A9bHsBrqXNAXjMCwo/edit?pli=1#)" document indicates that DB changes should have DB migrations. I added a link to this PR as an exclusion to that rule. In the future, when we have a DB version check in the installation/update process, the logic should be moved there and a DB migration generated for this functionality IMO.

Tested the PR by applying the migration. The new column is created successfully on MariaDB 10.2:

![image](https://user-images.githubusercontent.com/17739158/89718961-d1354200-d9c3-11ea-87aa-b36735adc209.png)

The "emails in time" widget on the dashboard also works as expected:

![image](https://user-images.githubusercontent.com/17739158/89718977-f5911e80-d9c3-11ea-819a-1d508338b3d8.png)

Only thing is that the line chart below is empty, but it was already empty prior to applying this PR. So I'm assuming that I just don't have enough data in my local instance for that graph to render.

![image](https://user-images.githubusercontent.com/17739158/89718989-10639300-d9c4-11ea-8215-e047b86cfdff.png)

kuzmany reviewed (dismissed):
I've tested it and works like describe. 
Column was created, indexed and query looks use it.

escopecz committed:
use correct minimal MariaDB version

escopecz commented:
@dennisameling good catch. No idea how I missed it. There is even a link to that page in the constant's comment. Fixed.
